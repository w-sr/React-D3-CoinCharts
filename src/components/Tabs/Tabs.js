/* eslint-disable react/jsx-no-bind */
import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Flex from "../Flex";
import { animation, border, color, fontSize, fontWeight } from "../../styles/constants";

const StyledTabs = styled(Flex)`
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium};
`;

const StyledTab = styled(Flex).attrs({
  color: props => (props.selected ? color.coincharts : color.coinchartsGray),
  border: props => (props.selected ? `1px solid ${color.coincharts}` : border.border),
})`
  color: ${props => props.color};
  height: inherit;
  margin: 0 0.5em;
  position: relative;
  transition: ${animation.default};

  &:first-child {
    margin-left: 0;
  }

  &:last-child {
    margin-right: 0;
  }

  &:focus {
    outline: none;
  }

  &::after {
    border-bottom: ${props => props.border};
    bottom: -1px;
    content: "";
    height: 1px;
    left: 0;
    position: absolute;
    width: 100%;
  }

  &:hover::after {
    border-bottom-color: ${props => props.color};
  }
`;

const Tabs = ({ options, onChange, selectedKey }) => (
  <StyledTabs role="tabpanel">
    {Object.keys(options).map(key => (
      <StyledTab
        align="center"
        selected={key === selectedKey}
        key={options[key].listKey}
        aria-labelledby={options[key].listKey}
        aria-selected={key === selectedKey}
        onClick={onChange.bind(null, key)}
        role="tab"
        tabIndex="-1"
      >
        {options[key].element}
      </StyledTab>
    ))}
  </StyledTabs>
);

Tabs.propTypes = {
  onChange: PropTypes.func.isRequired,
  options: PropTypes.shape({
    key: PropTypes.string,
  }).isRequired,
  selectedKey: PropTypes.string.isRequired,
};

export default Tabs;
