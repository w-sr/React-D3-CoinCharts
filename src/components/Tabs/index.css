.Tabs {
  display: flex;
  color: #7d95b6;
  flex-direction: row;
  font-size: 14px;
  font-weight: 500;
  line-height: normal;
}

.Tabs-item {
  align-items: center;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  height: inherit;
  margin: 0 0.5em;
  position: relative;
  transition: all 0.25s ease;
}

.Tabs-item:first-child {
  margin-left: 0;
}

.Tabs-item:last-child {
  margin-right: 0;
}

.Tabs-item:focus {
  outline: none;
}

.Tabs-item:hover::after {
  border-bottom-color: #7d95b6;
}

.Tabs-item::after {
  bottom: -1px;
  border-bottom: 1px solid transparent;
  content: '';
  height: 1px;
  left: 0;
  position: absolute;
  width: 100%;
}

.Tabs-item.selected {
  color: #0667d0;
}

.Tabs-item.selected::after {
  bottom: -1px;
  border-bottom: 1px solid #0667d0;
  content: '';
  height: 1px;
  left: 0;
  position: absolute;
  width: 100%;
}
