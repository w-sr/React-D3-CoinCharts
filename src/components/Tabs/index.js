import React from 'react';
import PropTypes from 'prop-types';

import './index.css';

const Tabs = ({ children, keys, onChange, selectedIndex }) => (
  <div className="Tabs" role="tabpanel">
    {
      children.map((child, index) => {
        const isSelectedClass = (index === selectedIndex) && 'selected';
        return (
          <div
            aria-labelledby={keys[index]}
            className={`Tabs-item ${isSelectedClass}`}
            key={keys[index]}
            onClick={() => onChange(index)}
            role="tab"
            tabIndex="-1"
          >
            {child}
          </div>
        );
      })
    }
  </div>
);

Tabs.propTypes = {
  children: PropTypes.arrayOf(PropTypes.element).isRequired,
  keys: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func.isRequired,
  selectedIndex: PropTypes.number.isRequired,
};

export default Tabs;
