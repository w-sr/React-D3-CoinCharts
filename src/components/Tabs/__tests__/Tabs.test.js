import React from "react";
import { render, Simulate } from "react-testing-library";

import Tabs from "../";

describe("<Tabs />", () => {
  it("renders without crashing", () => {
    const props = {
      options: {
        a: { listKey: "listKey_a", element: <div>a</div> },
        b: { listKey: "listKey_b", element: <div>b</div> },
      },
      onChange: () => {},
      selectedKey: "a",
    };

    const { container } = render(<Tabs {...props} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders selected tab correctly", () => {
    const selectedKey = "a";
    const props = {
      options: {
        a: { listKey: "listKey_a", element: <div>a</div> },
        b: { listKey: "listKey_b", element: <div>b</div> },
      },
      onChange: () => {},
      selectedKey,
    };

    const { container } = render(<Tabs {...props} />);
    expect(container.querySelector("div[aria-selected='true']")).not.toBeNull();
    expect(container.querySelector("div[aria-selected='true']").textContent).toEqual(selectedKey);
  });

  it("triggers `props.onChange` callback when tab is clicked", async () => {
    const initialSelectedKey = "a";
    const selectedKey = "b";
    const props = {
      options: {
        a: { listKey: "listKey_a", element: <div>a</div> },
        b: { listKey: "listKey_b", element: <div>b</div> },
      },
      onChange: jest.fn(),
      selectedKey: initialSelectedKey,
    };

    const { container } = render(<Tabs {...props} />);

    props.onChange.mockClear();
    Simulate.click(container.querySelectorAll("div[role='tab']")[1]);

    // Check if the 1st argument (of the 1st invocation) matches what we clicked on
    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange.mock.calls[0][0]).toEqual(selectedKey);
    expect(props.onChange.mock.calls[0][0]).not.toEqual(initialSelectedKey);
  });
});
