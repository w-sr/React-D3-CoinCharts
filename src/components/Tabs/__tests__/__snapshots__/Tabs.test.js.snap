// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> renders without crashing 1`] = `
<div
  class="sc-bwzfXH cqtslC sc-bdVaJa cAcgKX"
  role="tabpanel"
>
  <div
    class="sc-htpNat czjRRp sc-bdVaJa dyxdRy"
    color="#0667D0"
    aria-labelledby="listKey_a"
    aria-selected="true"
    role="tab"
    tabindex="-1"
  >
    <div>
      a
    </div>
  </div>
  <div
    class="sc-htpNat lgKbrQ sc-bdVaJa dyxdRy"
    color="#7D95B6"
    aria-labelledby="listKey_b"
    aria-selected="false"
    role="tab"
    tabindex="-1"
  >
    <div>
      b
    </div>
  </div>
</div>
`;
